CATS/Problem/TestsParser.pm

```diff
@@ -32,9 +33,66 @@ sub validate_test
         and return 'Both output file and standard solution';
     ($test->{points} // '0') =~ /^\d+$/
         or return 'Bad points';
+    if (my $error = $self->validate_by_formal($test)) {
+        return $error;
+    }
     undef;
 }
 
+sub validate_by_formal {
+    (my CATS::Problem::Parser $self, my $test) = @_;
+    my $formals = {
+        INPUT => $test->{input_validator_id} && $self->get_formal_src_by_id($test->{input_validator_id}),
+        OUTPUT => $test->{output_validator_id} && $self->get_formal_src_by_id($test->{output_validator_id})
+    };
+    my $files = {
+        INPUT => $test->{in_file},
+        OUTPUT => $test->{out_file}
+    };
+    my $should_validate_input = defined $files->{INPUT} ? $formals->{INPUT} : undef;
+    my $should_validate_output = (defined $files->{INPUT} || !defined $formals->{INPUT}) && defined $files->{OUTPUT} ?
+        $formals->{OUTPUT} : undef;
+    my $to_validate = {
+        INPUT => $should_validate_input,
+        OUTPUT => $should_validate_output
+    };
+    if ($should_validate_input || $should_validate_output) {
+        my $error = CATS::Formal::Formal::validate($to_validate, $files);
+        $error && return $error;
+    }
+    my $should_check_input_syntax = defined $formals->{INPUT} != defined $to_validate->{INPUT};
+    my $should_check_output_syntax = defined $formals->{OUTPUT} != defined $to_validate->{OUTPUT};
+    if ($should_check_input_syntax || $should_check_output_syntax) {
+        my $error = CATS::Formal::Formal::check_syntax(%$formals);
+        $error && return $error;
+    }
+    return undef;
+}
+
+sub find_first {
+    my ($code, @arr) = @_;
+    for (@arr) {
+        return $_ if $code->($_);
+    }
+    return undef;
+}
+
+sub get_formal_src_by_id
+{
+    (my CATS::Problem::Parser $self, my $id) = @_;
```

Comment:
We did not need to search by id before, but this is not a good way to do it. Create a hash of `id => member` instead.